### PR TITLE
[5.6] Return the collection iterator from query builder and relations

### PR DESCRIPTION
**Pull request by luiz-brandao**
I always forget to call `get()` to get the results of a query. Maybe because `all()`, `first()`, `find()` and similar methods already return a collection, or maybe because I'm just a noob :-) . Anyway, I'm not sure there is a reason for _not_ having this feature, but I always thought it would be cool to have an implicit `get()` when trying to iterate through a query or relation.  This PR allows you to do just that:

`foreach(User::someQueryScope() as $user){ ... }`

`foreach(User::posts() as $post){ ... }`

### CODE REVIEW

**Comment by sisve** [comment]
I believe you forgot to commit the tests.

**Comment by Dylan-DPC** [comment]
I remember this being proposed somewhere and many people didn't like it. Can't find the link.

**Comment by 36864** [comment]
I don't quite get you reasoning here.

>Maybe because all(), first(), find() and similar methods already return a collection,

So, you're fine with having to use `first()` and `find()`, but not `get()`?

That aside, I don't like this because this would mean that if I forget to specify what I want from the query builder, I'll have a `get()` when I might have wanted a `first()` or something else entirely. It's easier to detect that you've forgotten to finish your query by checking if you have an instance of query builder instead of collection than to detect that you're getting a collection instead of a different collection that you wanted.

For example, if I forget to call find() in the following snippet, right now nothing happens and I can easily determine why. With this change, every user would receive points and I'd have to explain to my boss why every user just got bonus points, and then we'd probably have to explain to a lot of users why the points they just got are being taken away.
```
$special_users = App\User::where('active', 1)->find([1, 2]);
foreach($special_users as $user) {
  $user->givePoints(10);
}
```

In short, I do not expect an instance of QueryBuilder to behave as an instance of Collection, and I think this would cause more problems than it solves.

**Comment by luiz-brandao** [comment]
@36864 My reasoning being that we don't need to call `get()` in all of those methods; they implicitly build a query and return either a collection or a model. 

I understand your point, but there are many things that can go wrong when you improperly build a query. I'm not sure it would cause more problems than it solves...

**Comment by Dylan-DPC** [comment]
What if you have a query object and later you want to dynamically manipulate the query?

**Comment by luiz-brandao** [comment]
@Dylan-DPC why would you try to iterate over it then?

**Comment by luiz-brandao** [comment]
I have to admit that this morning after @36864's comment I got a fresh view of how `get()` fits in the scheme of things... although I still think it would make the interface cleaner to get an iterator from the query builder.

**Comment by 36864** [comment]
@luiz-brandao 

I've always thought of `get()`, `find()`, etc. as query terminator methods. You call them when you're finished defining your query and you get a collection. Attempting to iterate through an unfinished query just does not make sense to me, and that's exactly what iterating over a QueryBuilder instance is. You're trying to iterate over a query before you've declared that you've finished building it.

Forgetfulness aside, I can see a case where you're building dynamic queries and conditionally calling `get()`, `first()`, or passing the builder along to another function that will terminate it. 
Something like `$results = finishQuery($my_query_builder));`. There may be some logic in that `finishQuery` method that will ultimately either call `get()`, or `find()` on the builder object and just return it. There may also be a case where nothing is called on the object before returning it, and you can safely `foreach($results as $result)` regardless of what happened to it. You will either iterate through the collection you expect to get, or do nothing. 
With this change, you will first have to check the type of `$results` before using it.

What you call a cleaner interface I consider a laziness enabler. Don't bother terminating your queries, the system will just finish them for you. That's a step towards a framework that would let you do something like `foreach(User::where('John') as $john)` which may sound like a nice shortcut for a developer who just wants to find John in the database, but is a nightmare to maintain years down the line when someone else is trying to fix an edge case bug in your software.

**Comment by luiz-brandao** [comment]
@36864 It makes sense and I like the idea of a query finisher. Another goal that I forgot to mention is the ability make the query lazy and just run it when I actually want the results. 

Maybe that's not the job of the query builder, but that would make things simpler again. Probably a lazy() method would be an interesting addition?

**Comment by 36864** [comment]
>Another goal that I forgot to mention is the ability to make the query lazy and just run it when I actually want the results.

That is precisely what the query terminators do. You can hold a QueryBuilder instance indefinitely, and you will only get the results when you specifically call `->get()`, `->find()`, `->first()`, or another getter function.

**Comment by luiz-brandao** [comment]
Of course :) but I had a more complex scenario in mind where I would like to avoid calling `get()` more than once for the same query in case it got loaded already, but it seems more like a domain problem and my PR would not solve for that anyway.

**Comment by sisve** [comment]
Two days later and this issue still lacks tests.

**Comment by luiz-brandao** [comment]
There you go!

**Comment by timacdonald** [comment]
This has already been suggested - with a fair bit of discussion that might add to this thread: https://github.com/laravel/internals/issues/563

I'd say thats what you had seen previously @Dylan-DPC

**Comment by luiz-brandao** [comment]
@timacdonald thanks for that, I searched for previous PR's relating to this but forgot to check the internals!

I like the query cache and immutable query ideas, I think they are worth exploring!

**Comment by deleugpn** [comment]
I'm with 36864 on this. I hate forgetting to call `get` as much as the next developer, but forgetting `first` or `find` would be a nightmare in this new scenario.

**Comment by mateusjatenee** [comment]
@deleugpn perhaps I understood everything wrong, but why would `first` or `find` change? From what I saw the only difference would be when the Builder is iterated over. Forgetting `find` and then calling a property would still give a `Undefined property` notice, wouldn't it?

**Comment by luiz-brandao** [comment]
I understand the fear of having an expensive query ran by mistake, but think about it... wouldn't it be the same type of mistake if I called `get` instead of `find`? Also, if what I really want is a `first`, wouldn't it be an error to iterate over it? 

Remember that the query will only be executed when an Iterable is expected. So if you put your "unfinished" query (the one expected to return a single model) in a foreach, is it really the fault of the feature?

**Comment by 36864** [comment]
The problem isn't running an expensive query by mistake, it's running the wrong query because something went wrong and instead of doing nothing, iterating through a builder would now do a query that hasn't been finished.
I'm much less worried with wasting a few milliseconds than I am with running a query that hasn't had all the constraints attached.

Forgetting to finish your query is not the same type of mistake as finishing your query wrongly.

Iterating over the builder object can happen for two reasons: The first and most common is forgetting to terminate your query, the second is you have complex query logic and handle it by conditionally adding constraints to a base query. While your solution fixes the first scenario, it causes severe failure on the second scenario. 
Since builders return an empty iterator, right now it's perfectly valid to build your query, conditionally call `get()`/`find()` and pass the result into a `foreach` loop. If the builder was terminated, you get results, if it wasn't for any reason, then nothing really happens.

I would much prefer that trying to iterate over a builder object just threw an exception or at the very least issued a warning, but I'll take silent failure over unintended query execution.

Also, iterating through a Model instance could be a valid way to iterate through its attributes, so a `foreach` on the result of a `first()` isn't necessarily an error.

**Comment by taylorotwell** [comment]
I don't think I'll be adding this for now.

**Comment by luiz-brandao** [comment]
I agree with most points raised so far to some extent. I may release this as a package so people interested may try it and check the impact it will have in their codebase. Thanks for the good feedback y'all :)
